core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java

```diff
@@ -425,38 +425,38 @@ private static void sqlError(
         );
     }
 
-    private void compileQuery(
+    private void compileAndExecuteQuery(
             JsonQueryProcessorState state
     ) throws PeerDisconnectedException, PeerIsSlowToReadException, QueryPausedException, SqlException {
         boolean recompileStale = true;
-        for (int retries = 0; recompileStale; retries++) {
-            try {
+        try (SqlCompiler compiler = engine.getSqlCompiler()) {
+            for (int retries = 0; recompileStale; retries++) {
                 final long nanos = nanosecondClock.getTicks();
-                try (SqlCompiler compiler = engine.getSqlCompiler()) {
-                    final CompiledQuery cc = compiler.compile(state.getQuery(), sqlExecutionContext);
-                    sqlExecutionContext.storeTelemetry(cc.getType(), TelemetryOrigin.HTTP_JSON);
-                    state.setCompilerNanos(nanosecondClock.getTicks() - nanos);
-                    state.setQueryType(cc.getType());
-                    // todo: reconsider whether we need to keep the SqlCompiler instance open while executing the query
-                    // the problem is the each instance of the compiler has just a single instance of the CompilerQuery object.
-                    // the CompilerQuery is used as a flyweight(?) and we cannot return the SqlCompiler instance to the pool
-                    // until we extract the result from the CompilerQuery.
+                final CompiledQuery cc = compiler.compile(state.getQuery(), sqlExecutionContext);
+                sqlExecutionContext.storeTelemetry(cc.getType(), TelemetryOrigin.HTTP_JSON);
+                state.setCompilerNanos(nanosecondClock.getTicks() - nanos);
+                state.setQueryType(cc.getType());
+                // todo: reconsider whether we need to keep the SqlCompiler instance open while executing the query
```

Comment:
nit: maybe it's time to remove this TODO? A retained compiler instance doesn't sound like a huge problem considering that we need it for each recompilation attempt.

Reply:
the problem is that the compiler is retained even during the query execution. I am planning to fix this, but not in this PR.

Reply:
I forgot that we need the `CompiledQuery` object in a few types of statements such as INSERT or UPDATE. In other types, we don't need it: more than that, query execution may need to be suspended due to Cold Storage or simply because the send buffer is full and we need to wait for the client to receive the result set chunk.

So, in theory we could retain `CompiledQuery` only in a few query types. On the other hand, we make sure to size the compiler pool with the number of threads * 2, so retaining a compiler shouldn't lead to thread starvation.

Reply:
yeah, I agree. still, I would prefer to have it in a separate PR. I would like to have this this is strictly about the `TableReferenceOutOfDateException` if possible.